### PR TITLE
Add support for functions

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR optimize the execution of functions provided to configure_commands or configure_api by executing them direclty within the Rest API process. 

Fixes #\<issue_number>

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: src/lightning_app/api/http_methods.py [hunk 1]
```diff
@@ -29,61 +29,82 @@ def __init__(self, route: str, method: Callable, method_name: Optional[str] = No
...
+
+            route(self.route, **self.kwargs)(handle_request_method)
+        else:
+            # 1: Create a proxy function with the signature of the wrapped method.
+            fn = deepcopy(_signature_proxy_function)
+            fn.__annotations__ = self.method_annotations
+            fn.__name__ = self.method_name
+            setattr(fn, "__signature__", self.method_signature)
+
+            request_cls = CommandRequest if self.route.startswith("/command/") else APIRequest
```
Comment (awaelchli) [inline src/lightning_app/api/http_methods.py:71]: all this code is the same in both branches of the if-else

**Review by awaelchli — APPROVED** [review]

File: src/lightning_app/api/http_methods.py (hunk 1 above)
Comment (manskx) [reply src/lightning_app/api/http_methods.py:71]: I think this if-else needs to be revisited.

File: src/lightning_app/api/http_methods.py [hunk 2]
```diff
@@ -29,61 +29,82 @@ def __init__(self, route: str, method: Callable, method_name: Optional[str] = No
...
-                        id=request_id,
+        if not self.attached_to_flow:
+            fn = deepcopy(_signature_proxy_function)
+            fn.__annotations__ = self.method_annotations
+            fn.__name__ = self.method_name
+            setattr(fn, "__signature__", self.method_signature)
+
+            request_cls = CommandRequest if self.route.startswith("/command/") else APIRequest
+
+            # 3: Define the request handler.
```
Comment (manskx) [inline src/lightning_app/api/http_methods.py:58]: The comments are not in the correct order.

File: tests/tests_app_examples/test_commands_and_api.py [hunk 3]
```diff
@@ -45,3 +45,7 @@ def test_commands_and_api_example_cloud() -> None:
                 if "['this', 'is', 'awesome']" in log:
                     has_logs = True
             sleep(1)
+
+        # 7: Send a request to the Rest API directly.
+        resp = requests.get(base_url + "/pure_function")
+        assert resp.status_code == 200, resp.json()
```
Comment (manskx) [inline tests/tests_app_examples/test_commands_and_api.py:51]: Could it test the response as well ? `"Hello World !"`

**Review by manskx — APPROVED** [review]
Looks good, I left a few comments.

File: src/lightning_app/api/http_methods.py (hunk 1 above)
Comment (tchaton) [reply src/lightning_app/api/http_methods.py]: Done !

File: src/lightning_app/api/http_methods.py (hunk 2 above)
Comment (tchaton) [reply src/lightning_app/api/http_methods.py:58]: Done !

File: tests/tests_app_examples/test_commands_and_api.py (hunk 3 above)
Comment (tchaton) [reply tests/tests_app_examples/test_commands_and_api.py]: Done !

**Review by lantiga — APPROVED** [review]
LGTM

**Pull request merged**